### ISSUE TITLE
[Question] Namespace-scoped kubectl access

### ISSUE THREAD

**Issue by TeamDman** [Azure/AKS#3556]
# Problem

I am setting up a cluster to be used by multiple independent client groups.  
The clients should be able to execute kubectl commands for resources in their namespace, but NOT other namespaces.

I have given my test account the "Azure Kubernetes Service Cluster User Role" role, and I am able to successfully run `az aks get-credentials`, but I am ALSO ABLE TO VIEW SECRETS without assigning "Azure Kubernetes Service RBAC Reader".

Is the AKS RBAC Reader role only for the portal then?

```hcl
role_definition_name = "Azure Kubernetes Service RBAC Reader"
scope                = "${cluster_id}/namespaces/${namespace}"
```

--- 

I removed the "Azure Kubernetes Service Cluster User Role" and was, as expected, denied when trying to get the credentials again

```
$ az aks get-credentials --resource-group SCED-NPRD-CACN-TC3oE-Shared-Cluster-RGP --name tc3oesharedaksnprd --subscription SCED-NPRD
(AuthorizationFailed) The client 'abc@def' with object id 'asd' does not have authorization to perform action 'Microsoft.ContainerService/managedClusters/listClusterUserCredential/action' over scope '/subscriptions/123/resourceGroups/SCED-NPRD-CACN-TC3oE-Shared-Cluster-RGP/providers/Microsoft.ContainerService/managedClusters/tc3oesharedaksnprd' or the scope is invalid. If access was recently granted, please refresh your credentials.
Code: AuthorizationFailed
Message: The client 'abc@def' with object id 'asd' does not have authorization to perform action 'Microsoft.ContainerService/managedClusters/listClusterUserCredential/action' over scope '/subscriptions/123/resourceGroups/SCED-NPRD-CACN-TC3oE-Shared-Cluster-RGP/providers/Microsoft.ContainerService/managedClusters/tc3oesharedaksnprd' or the scope is invalid. If access was recently granted, please refresh your credentials.
```

However, I am still able to use kubectl to view secrets until I delete the kubeconfig.

# Questions 

- Is it impossible to remove cluster access once the kubectl config is pulled?
- Does `az aks get-credentials` always give cluster-wide access?
- Is it possible to give access to ONLY a single namespace for kubectl and the portal, using only Azure RBAC and no additional K8s resources?

# Details

According to Azure docs, this should be straight forward with Azure RBAC and without requiring Kubernetes role definitions and assignments [\[ref\]](https://learn.microsoft.com/en-us/azure/aks/manage-azure-rbac#create-role-assignments-for-users-to-access-the-cluster).  
> Azure Kubernetes Service RBAC Reader
> Allows read-only access to see most objects in a namespace. It doesn't allow viewing roles or role bindings. This role doesn't allow viewing Secrets, since reading the contents of Secrets enables access to ServiceAccount credentials in the namespace, which would allow API access as any ServiceAccount in the namespace (a form of privilege escalation).

In order for our clients to be able to call `az aks get-credentials`, they need the "Azure Kubernetes Service Cluster User Role" role 
[\[ref\]](https://stackoverflow.com/a/69719593/11141271) [\[ref\]](https://learn.microsoft.com/en-us/azure/aks/control-kubeconfig-access#available-cluster-roles-permissions) [\[ref\]](https://techcommunity.microsoft.com/t5/fasttrack-for-azure/azure-kubernetes-service-rbac-options-in-practice/ba-p/3684275) [\[ref\]](https://learn.microsoft.com/en-us/azure/aks/concepts-identity#summary).  

> Azure Kubernetes Service Cluster User Role  
> - Allows access to Microsoft.ContainerService/managedClusters/listClusterUserCredential/action API call. This API call lists the cluster user credentials.  
> - Downloads kubeconfig for clusterUser role.

The [role definition](https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#azure-kubernetes-service-cluster-user-role)
```json
{
  "assignableScopes": [
    "/"
  ],
  "description": "List cluster user credential action.",
  "id": "/subscriptions/{subscriptionId}/providers/Microsoft.Authorization/roleDefinitions/4abbcc35-e782-43d8-92c5-2d3f1bd2253f",
  "name": "4abbcc35-e782-43d8-92c5-2d3f1bd2253f",
  "permissions": [
    {
      "actions": [
        "Microsoft.ContainerService/managedClusters/listClusterUserCredential/action",
        "Microsoft.ContainerService/managedClusters/read"
      ],
      "notActions": [],
      "dataActions": [],
      "notDataActions": []
    }
  ],
  "roleName": "Azure Kubernetes Service Cluster User Role",
  "roleType": "BuiltInRole",
  "type": "Microsoft.Authorization/roleDefinitions"
}
```

Then it should be as straight forward as applying the "Azure Kubernetes Service RBAC Reader" role to the proper scope, including the namespace [\[ref\]](https://learn.microsoft.com/en-us/azure/aks/manage-azure-rbac#create-role-assignments-for-users-to-access-the-cluster)

> Allows read-only access to see most objects in a namespace. It doesn't allow viewing roles or role bindings. This role doesn't allow viewing Secrets, since reading the contents of Secrets enables access to ServiceAccount credentials in the namespace, which would allow API access as any ServiceAccount in the namespace (a form of privilege escalation).

From what I can tell, I've enabled the RBAC stuff correctly.

Custer json excerpt:

```json
        "enableRBAC": true,
        "aadProfile": {
            "managed": true,
            "adminGroupObjectIDs": [
                "redacted"
            ],
            "adminUsers": null,
            "enableAzureRBAC": true,
            "tenantID": "redacted"
        },
```

I double checked, and the specified admin group does NOT contain the account I am testing with.

**Comment by TeamDman** [comment]
Seems like the `kubelogin convert-kubeconfig` (or something else?) was persisting my admin session instead of fully logging out when I delete the `~/.kube/config` file. Looking at the log workspace connected to the cluster revealed that my commands were still being executed as the admin account even after running `az aks get-credentials`

Deleting the kubeconfig AND running the kubelogin command to remove tokens seems to have cleared up the issue.

I am now able to login as the non-admin account and I am appropriately denied when invoking kubectl commands that reach beyond the access given by the Kubernetes RoleBindings.

---

TL;DR I was still logged in to the cluster as admin, so of course I could access everything :P
Double check which user you are using logs I guess.
Don't log in to your cluster using a library computer because apparently it's easy to not fully log out properly.
